data/pools/transactionPool.go

```diff
@@ -179,6 +179,44 @@ func (pool *TransactionPool) Remember(t transactions.SignedTxn) error {
 	defer pool.mu.Unlock()
 
 	err := pool.test(t)
+	// Check for a race condition when a new block was added
+	// to our ledger, but hasn't propagated to OnNewBlock()
+	// yet.
+	switch errt := err.(type) {
+	case transactions.TxnDeadError:
+		if errt.Round >= errt.FirstValid {
+			// This is not an error due to the round being too old.
+			break
+		}
+
+		if pool.pendingBlockEvaluator == nil {
+			// We don't have a block evaluator at all; something
+			// else went wrong, and this isn't the place to fix
+			// it.  (The next OnNewBlock callback might recover.)
+			break
```

Comment:
You might want to change the returned err object here; it's no longer a dead transaction ( i.e. error 400 Bad Request) but rather ( 500 Internal Server Error )

Reply:
I think this is unreachable anyway, because `TransactionPool.test()` never returns `TxnDeadError` if `pool.pendingBlockEvaluator == nil`.  It just didn't seem right to dereference the possibly-nil `pool.pendingBlockEvaluator` in this block of code without explicitly checking for nil, even if right now the surrounding code is such that it can't happen.

Reply:
if that's the case, could you just add a comment explaining that ?